### PR TITLE
fix: pad prefill kernel to uncached suffix length with prefix caching

### PR DESCRIPTION
**Pull request by ryankert01**
## Intuition

prefix sharing is not doing great at my benchmark.

## Benchmark

64-prompt burst, 8K shared prefix + 256 unique suffix + 256 output

### Before patch (prefix caching ON vs OFF)

| Metric | ON (caching) | OFF | Ratio |
|---|---:|---:|:---|
| Mean TTFT | 259 s | 184 s | 0.71× (ON slower) |
| p99 TTFT | 363 s | 374 s | ~equal |
| Goodput | 40.7 tok/s | 41.9 tok/s | 0.97× |
| Wall time | 403 s | 391 s | ~equal |
| Hit rate | 95.1% | 0% | — |

95% hit rate but effectively zero performance gain — the kernel was doing full-prompt compute regardless.

### After patch (prefix caching ON vs OFF)

| Metric | ON (patched) | OFF | Speedup |
|---|---:|---:|:---|
| Wall time | 103.1 s | 391.5 s | **3.80×** |
| Mean TTFT | 37.3 s | 184.2 s | **4.94×** |
| p99 TTFT | 62.8 s | 374.6 s | **5.97×** |
| Goodput | 158.9 tok/s | 41.8 tok/s | **3.80×** |
| Hit rate | 95.1% | 0% | — |

Patch delivers **3.9× wall-time improvement** and **4.9× mean TTFT improvement** vs the unpatched ON run on the same workload.

### CODE REVIEW

**Comment by yieldthought** [comment]
Codex review:

[models/tt_transformers/tt/generator.py:546](https://github.com/tenstorrent/tt-metal/blob/02923f4ad8641320ba6a4673f9843bc9326f7846/models/tt_transformers/tt/generator.py#L546) - This now passes suffix-padded prefill_seq_lens into the row-sharded batched prefill path before the new no-prefix-cache guard is applied. That callee still treats prefill_seq_lens as padding for the full prompt: [models/demos/gpt_oss/tt/model.py:639](https://github.com/tenstorrent/tt-metal/blob/02923f4ad8641320ba6a4673f9843bc9326f7846/models/demos/gpt_oss/tt/model.py#L639) slices tokens[:plen], then [models/demos/gpt_oss/tt/model.py:640](https://github.com/tenstorrent/tt-metal/blob/02923f4ad8641320ba6a4673f9843bc9326f7846/models/demos/gpt_oss/tt/model.py#L640) pads with max_padded_len - plen. For a GPT-OSS row-sharded vLLM batch (users_row_sharded, harmony prompt, sampling_params set), plen is the full prompt length while max_padded_len is now only the uncached suffix padding; e.g. an 8K cached prefix plus a 256-token suffix gives max_padded_len=1024 and plen around 8448, so torch.zeros gets a negative dimension before prefill runs. Suggestion: add the same all(n == 0 for n in num_cached_per_user) guard to the row-sharded branch, or plumb start_pos into row_sharded_batched_prefill and slice/pad the uncached suffix there.

File: models/tt_transformers/tt/generator.py [hunk 1]
```diff
@@ -513,7 +513,18 @@ def prefill_forward_text(
...
+        # kernel (see prefill_ids construction below), so padding by full seq_len makes
+        # the kernel run far more compute than necessary for high-hit-rate workloads.
+        # int() cast handles numpy.int64 from vLLM (bit_length() needs a Python int).
+        num_cached_per_user = (
+            [int(n) for n in start_pos] if start_pos is not None else [0] * len(prompt_lens)
+        )
+        prefill_seq_lens = [
+            get_padded_prefill_len(int(seq_len) - num_cached)
+            for seq_len, num_cached in zip(prompt_lens, num_cached_per_user)
+        ]
```
[the remark this answers is not in the document]
Comment (ryankert01) [reply models/tt_transformers/tt/generator.py:527]: make sense

**Comment by mtairum** [comment]
Dispatched **all-model-tests** for this PR (tiers 1+2+3, e2e + unit, all SKUs):

https://github.com/tenstorrent/tt-metal/actions/runs/25814861441

Branch: `ci-test/pr-44194` (temp branch in this repo pointing at `b231ea6c` — same commit as this PR head, since dispatch can't run directly on a fork ref). Safe to delete once the run finishes.

**Review by mtairum — CHANGES_REQUESTED** [review]
This looks ok, but we'll have to wait for the result of 
`all-model-tests` (tiers 1+2+3, e2e + unit, all SKUs):

https://github.com/tenstorrent/tt-metal/actions/runs/25814861441

When this is done, we can compare to the status on main, and if there's no regressions, I'll re-review.

**Comment by ryankert01** [comment]
Those CI tests failed seems unrelated to this change. wdyt? @mtairum

**Review by mtairum — APPROVED** [review]

**Comment by mtairum** [comment]
Rebased and re-running the PR checks. Will be merged once they pass. Thank you.

**Comment by ryankert01** [comment]
Thanks for the review!
